source/fuzz/transformation_merge_function_returns.cpp

```diff
@@ -0,0 +1,646 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_merge_function_returns.h"
+
+#include "source/fuzz/comparator_deep_blocks_first.h"
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationMergeFunctionReturns::TransformationMergeFunctionReturns(
+    const protobufs::TransformationMergeFunctionReturns& message)
+    : message_(message) {}
+
+TransformationMergeFunctionReturns::TransformationMergeFunctionReturns(
+    uint32_t function_id, uint32_t outer_header_id, uint32_t outer_return_id,
+    uint32_t return_val_id, uint32_t any_returnable_val_id,
+    const std::vector<protobufs::ReturnMergingInfo>& returns_merging_info) {
+  message_.set_function_id(function_id);
+  message_.set_outer_header_id(outer_header_id);
+  message_.set_outer_return_id(outer_return_id);
+  message_.set_return_val_id(return_val_id);
+  message_.set_any_returnable_val_id(any_returnable_val_id);
+  for (const auto& return_merging_info : returns_merging_info) {
+    *message_.add_return_merging_info() = return_merging_info;
+  }
+}
+
+bool TransformationMergeFunctionReturns::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto function = ir_context->GetFunction(message_.function_id());
+  // The function must exist.
+  if (!function) {
+    return false;
+  }
+
+  // The entry block must end in an unconditional branch.
+  if (function->entry()->terminator()->opcode() != SpvOpBranch) {
+    return false;
+  }
+
+  // If the function has a non-void return type,
+  // |message_.any_returnable_val_id| must exist, have the same type as the
+  // return type of the function and be available at the end of the entry block.
+  auto function_type = ir_context->get_type_mgr()->GetType(function->type_id());
+  assert(function_type && "The function type should always exist.");
+
+  // Get a map from the types for which ids are available at the end of the
+  // entry block to one of the ids with that type. We compute this here to avoid
+  // potentially doing it multiple times later on.
+  auto types_to_available_ids =
+      GetTypesToIdAvailableAfterEntryBlock(ir_context);
```

Comment:
I would declare this (and it's comment) before the comment about non-void return type, as it interrupts the flow here.  This is generally useful throughout the function; it reads as if it's going to be used for the checking of message_.any_returnable_id.

Reply:
I moved it up